src/sciline/pipeline.py

```diff
@@ -74,14 +94,139 @@ def _bind_free_typevars(tp: TypeVar | Key, bound: Dict[TypeVar, Key]) -> Key:
         return tp
 
 
+def _find_all_paths(
+    graph: Mapping[T, Collection[T]], start: T, end: T
+) -> List[List[T]]:
+    """Find all paths from start to end in a DAG."""
+    if start == end:
+        return [[start]]
+    if start not in graph:
+        return []
+    paths = []
+    for node in graph[start]:
+        for path in _find_all_paths(graph, node, end):
+            paths.append([start] + path)
+    return paths
+
+
+def _find_nodes_in_paths(
+    graph: Mapping[T, Tuple[Callable[..., Any], Collection[T]]], start: T, end: T
+) -> List[T]:
+    """
+    Find all nodes that need to be duplicated since they depend on a value from a
+    param table.
```

Comment:
The docstring refers to a specific use case, not to what the function does.

Reply:
Yes, is that bad? It is an implementation detail.

Reply:
It confused my at first. Maybe just mention that this is the particular use case.